File: src/main/java/org/example/eshop/config/CorrelationIdFilter.java
package org.example.eshop.config;

import jakarta.servlet.Filter;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.example.eshop.service.CorrelationIdService;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Component
public class CorrelationIdFilter implements Filter {

    private final CorrelationIdService correlationIdService;

    public CorrelationIdFilter(CorrelationIdService correlationIdService) {
        this.correlationIdService = correlationIdService;
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws IOException, ServletException {
        HttpServletRequest httpRequest = (HttpServletRequest) request;
        HttpServletResponse httpResponse = (HttpServletResponse) response;
        try {
            String correlationId = httpRequest.getHeader(CorrelationIdService.CORRELATION_ID_HEADER);
            if (correlationId == null || correlationId.isBlank()) {
                correlationId = correlationIdService.generateCorrelationId();
            }
            correlationIdService.setCorrelationId(correlationId);
            httpResponse.setHeader(CorrelationIdService.CORRELATION_ID_HEADER, correlationId);
            chain.doFilter(request, response);
        } finally {
            correlationIdService.clearCorrelationId();
        }
    }
}

File: src/main/kotlin/org/example/eshop/config/CorrelationIdFilter.kt
package org.example.eshop.config

import org.example.eshop.service.CorrelationIdService
import org.springframework.stereotype.Component
import jakarta.servlet.*
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse

@Component
class CorrelationIdFilter(
    private val correlationIdService: CorrelationIdService
) : Filter {

    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        val httpRequest = request as HttpServletRequest
        val httpResponse = response as HttpServletResponse
        
        try {
            // Get correlation ID from header or generate new one
            val correlationId = httpRequest.getHeader(CorrelationIdService.CORRELATION_ID_HEADER)
                ?: correlationIdService.generateCorrelationId()
            
            // Set correlation ID in thread local
            correlationIdService.setCorrelationId(correlationId)
            
            // Add correlation ID to response header
            httpResponse.setHeader(CorrelationIdService.CORRELATION_ID_HEADER, correlationId)
            
            // Continue with the request
            chain.doFilter(request, response)
        } finally {
            // Clean up thread local to prevent memory leaks
            correlationIdService.clearCorrelationId()
        }
    }
}